[ADD] invoice_split: Add module to split customer invoices

luisg123v opened the pull request:
This module allows to split a customer invoice, so the customer may pay
a percentage of the invoice amount and, optionally,  create a new
invoice to pay the remaining amount later.

For more info, see the README file.

Dummy MR: https://git.vauxoo.com/absa/absa/merge_requests/75

JulioSerna commented:
@luisg123v @yaniaular 

Is there an UH related to this PR?

luisg123v commented:
Hi @JulioSerna  
This is related to the [task#27639](https://www.vauxoo.com/web#id=27639&view_type=form&model=project.task&action=464&active_id=9&menu_id=317)

luisg123v commented:
@yaniaular  
This is ready for review.

yaniaular commented on invoice_split/README.md:
```diff
@@ -0,0 +1,36 @@
+
+To use this module, you need to:
+
+1. Go to ani invoice
```
@luisg123v ani -> any
```suggestion
1. Go to any invoice
```

yaniaular commented on invoice_split/wizards/invoice_split.py:
```diff
@@ -0,0 +1,46 @@
+    invoice_id = fields.Many2one('account.invoice', String="Invoice to Split")
+    new_invoice_id = fields.Many2one('account.invoice')
+    percent = fields.Float(
+        string="Percent to Pay", default=50.0,
```
@luisg123v 
```suggestion
        string="Percent to Pay",
```

luisg123v replied:
Are you sure? If default value is not given, it takes one anyway, but it's zero.

luisg123v replied:
We could remove the default and raise a warning if the number if out of range, e.g. 0.

yaniaular commented on invoice_split/wizards/invoice_split.xml:
```diff
@@ -0,0 +1,36 @@
+        <field name="arch" type="xml">
+            <form string="Split Invoice">
+                <group>
+                    <field name="percent"/>
```
@luisg123v 
```suggestion
                    <field name="percent" required="1"/>
```

luisg123v replied:
This is a numeric field, `required` has no effect over it, because it always has a value. If you leave it empty, it will be set to zero.

luisg123v replied:
I added a validation to prevent this from being zero.

yaniaular replied:
Good. It's better

yaniaular commented:
@luisg123v Add translations please

yaniaular commented on invoice_split/wizards/invoice_split.py:
```diff
@@ -0,0 +1,46 @@
+        string="Create New Invoice", default=True,
+        help="Specifies whether a new invoice will be created in draft mode "
+        "with the remaining uncovered amount")
+
```
@luisg123v Check if the invoice is in draft state, something like this:
```suggestion
    @api.model
    def default_get(self, fields):
        rec = super(AccountInvoiceSplit, self).default_get(fields)
        active_ids = self.env.context.get('active_ids')
        active_model = self.env.context.get('active_model')

        # Check for selected invoices ids
        if not active_ids or active_model != 'account.invoice':
            return rec

        invoices = self.env['account.invoice'].browse(active_ids)

        # Check invoices are draft
        if any(invoice.state != 'draft' for invoice in invoices):
            raise UserError(_("You can only split draft invoices"))
```

luisg123v commented:
Done @yaniaular, translations were included.

luisg123v commented:
@yaniaular  
All comments were applied or answered.

Note: 
For now, this is red because of [this issue](https://github.com/Vauxoo/addons-vauxoo/pull/1304#discussion_r227633903)

luisg123v commented:
@yaniaular  
This is not red anymore.

yaniaular commented:
@luisg123v The taxes in the splited invoices are not recalculating, could you check if [this method](https://github.com/odoo/odoo/blob/12.0/addons/account/models/account_invoice.py#L732-L739) is called correctly in your method?  please

luisg123v committed:
[ADD] invoice_split: Add module to split customer invoices to pay a

This module allows to split a customer invoice, so the customer may pay
a percentage of the invoice amount and, optionally,  create a new
invoice to pay the remaining amount later.

For more info, see the README file.

luisg123v commented:
@yaniaular   done.

I also improved the README and docstrings to explain te case of invoicing three units and splitting by 50%, where 1.5 of any product  will be  invoiced, even if it's not possible in practice.

luisg123v commented:
We found this module to not be generic enough to be here (at least for now), hence I'm closing this.
